### DISCUSSION TITLE
A new and (almost) effective way to update routes and allow external access

### DISCUSSION THREAD

**Discussion by eduardolucioac** [okd-project/okd#739] [Q&A] [answered]
This is a procedure I tried to use to update all the routes of an already installed cluster to allow external access to it.

The procedure below seems very viable to me, but we have been having some problems so I would like your help to solve this big problem that has been consuming me for days... =|

-----------------------------------------

**PROCEDURES:**

I found all **_namespaces_** that have _routes_...

```
[root@okd4-services okd_bare_metal]# oc get routes --all-namespaces
NAMESPACE                  NAME                HOST/PORT                                                   PATH   SERVICES            PORT    TERMINATION            WILDCARD
openshift-authentication   oauth-openshift     oauth-openshift.apps.mbr.okd.local                                 oauth-openshift     6443    passthrough/Redirect   None
openshift-console          console             console-openshift-console.apps.mbr.okd.local                       console             https   reencrypt/Redirect     None
openshift-console          downloads           downloads-openshift-console.apps.mbr.okd.local                     downloads           http    edge/Redirect          None
openshift-ingress-canary   canary              canary-openshift-ingress-canary.apps.mbr.okd.local                 ingress-canary      8080    edge/Redirect          None
openshift-monitoring       alertmanager-main   alertmanager-main-openshift-monitoring.apps.mbr.okd.local          alertmanager-main   web     reencrypt/Redirect     None
openshift-monitoring       grafana             grafana-openshift-monitoring.apps.mbr.okd.local                    grafana             https   reencrypt/Redirect     None
openshift-monitoring       prometheus-k8s      prometheus-k8s-openshift-monitoring.apps.mbr.okd.local             prometheus-k8s      web     reencrypt/Redirect     None
openshift-monitoring       thanos-querier      thanos-querier-openshift-monitoring.apps.mbr.okd.local             thanos-querier      web     reencrypt/Redirect     None
```

I found all the resources related to each _namespace_ and took note of the _route_ configuration paths...

```
[root@okd4-services okd_bare_metal]# oc -n openshift-authentication get all
[...]
NAME                                       HOST/PORT                            PATH   SERVICES          PORT   TERMINATION            WILDCARD
route.route.openshift.io/oauth-openshift   oauth-openshift.apps.mbr.okd.local          oauth-openshift   6443   passthrough/Redirect   None
```

```
[root@okd4-services okd_bare_metal]# oc -n openshift-console get all
[...]
NAME                                 HOST/PORT                                        PATH   SERVICES    PORT    TERMINATION          WILDCARD
route.route.openshift.io/console     console-openshift-console.apps.mbr.okd.local            console     https   reencrypt/Redirect   None
route.route.openshift.io/downloads   downloads-openshift-console.apps.mbr.okd.local          downloads   http    edge/Redirect        None
```

```
[root@okd4-services okd_bare_metal]# oc -n openshift-ingress-canary get all
[...]
NAME                              HOST/PORT                                            PATH   SERVICES         PORT   TERMINATION     WILDCARD
route.route.openshift.io/canary   canary-openshift-ingress-canary.apps.mbr.okd.local          ingress-canary   8080   edge/Redirect   None
```

```
[root@okd4-services okd_bare_metal]# oc -n openshift-monitoring get all
[...]
NAME                                         HOST/PORT                                                   PATH   SERVICES            PORT    TERMINATION          WILDCARD
route.route.openshift.io/alertmanager-main   alertmanager-main-openshift-monitoring.apps.mbr.okd.local          alertmanager-main   web     reencrypt/Redirect   None
route.route.openshift.io/grafana             grafana-openshift-monitoring.apps.mbr.okd.local                    grafana             https   reencrypt/Redirect   None
route.route.openshift.io/prometheus-k8s      prometheus-k8s-openshift-monitoring.apps.mbr.okd.local             prometheus-k8s      web     reencrypt/Redirect   None
route.route.openshift.io/thanos-querier      thanos-querier-openshift-monitoring.apps.mbr.okd.local             thanos-querier      web     reencrypt/Redirect   None
```

In each route I updated the _host_ property to a new domain...

```
oc edit -n openshift-console route.route.openshift.io/console
oc edit -n openshift-console route.route.openshift.io/downloads
oc edit -n openshift-ingress-canary route.route.openshift.io/canary
oc edit -n openshift-monitoring route.route.openshift.io/alertmanager-main
oc edit -n openshift-monitoring route.route.openshift.io/grafana
oc edit -n openshift-monitoring route.route.openshift.io/prometheus-k8s
oc edit -n openshift-monitoring route.route.openshift.io/thanos-querier
```

In other words, I modified something like this...

```
apiVersion: route.openshift.io/v1
kind: Route
metadata:
[...]
spec:
  host: route-name.apps.mbr.okd.local
[...]
status:
[...]
    host: route-name.apps.mbr.okd.local
[...]
```

... to something like this...

```
apiVersion: route.openshift.io/v1
kind: Route
metadata:
[...]
spec:
  host: route-name.apps.mbr.mydomain.net
[...]
status:
[...]
    host: route-name.apps.mbr.mydomain.net
[...]
```

In the case of the _route_ `route.route.openshift.io/oauth-openshift` I needed to modify the _ingress_...

```
oc edit ingress.config.openshift.io
```

... modifing the _domain_ property from something like this...

```
apiVersion: config.openshift.io/v1
kind: Ingress
metadata:
[...]
spec:
  domain: apps.mbr.okd.local
[...]
```

... to something like this...

```
apiVersion: config.openshift.io/v1
kind: Ingress
metadata:
[...]
spec:
  domain: apps.mbr.mydomain.net
[...]
```

After doing the above procedures all my routes were updated to the new domain...

```
[root@okd4-services okd_bare_metal]# oc get routes --all-namespaces
NAMESPACE                  NAME                HOST/PORT                                                     PATH   SERVICES            PORT    TERMINATION            WILDCARD
openshift-authentication   oauth-openshift     oauth-openshift.apps.mbr.mydomain.net                                 oauth-openshift     6443    passthrough/Redirect   None
openshift-console          console             console-openshift-console.apps.mbr.mydomain.net                       console             https   reencrypt/Redirect     None
openshift-console          downloads           downloads-openshift-console.apps.mbr.mydomain.net                     downloads           http    edge/Redirect          None
openshift-ingress-canary   canary              canary-openshift-ingress-canary.apps.mbr.mydomain.net                 ingress-canary      8080    edge/Redirect          None
openshift-monitoring       alertmanager-main   alertmanager-main-openshift-monitoring.apps.mbr.mydomain.net          alertmanager-main   web     reencrypt/Redirect     None
openshift-monitoring       grafana             grafana-openshift-monitoring.apps.mbr.mydomain.net                    grafana             https   reencrypt/Redirect     None
openshift-monitoring       prometheus-k8s      prometheus-k8s-openshift-monitoring.apps.mbr.mydomain.net             prometheus-k8s      web     reencrypt/Redirect     None
openshift-monitoring       thanos-querier      thanos-querier-openshift-monitoring.apps.mbr.mydomain.net             thanos-querier      web     reencrypt/Redirect     None
```

-----------------------------------------

**PROBLEMS:**

However the following new problems arose...

I can access the new _**Web Console**_ route...

`https://console-openshift-console.apps.mbr.mydomain.net/`

... and it redirects to the route below (old domain)...

`https://oauth-openshift.apps.mbr.okd.local/oauth/authorize?client_id=console&redirect_uri=https%3A%2F%2Fconsole-openshift-console.apps.mbr.mydomain.net%2Fauth%2Fcallback&response_type=code&scope=user%3Afull&state=3ba1134a`

![Screenshot_20210703_012238](https://user-images.githubusercontent.com/6520357/124343508-527c6100-dba2-11eb-918f-5dca677b8415.png)

... but we can use the new route...

`https://oauth-openshift.apps.mbr.mydomain.net/oauth/authorize?client_id=console&redirect_uri=https%3A%2F%2Fconsole-openshift-console.apps.mbr.mydomain.net%2Fauth%2Fcallback&response_type=code&scope=user%3Afull&state=3ba1134a`

...so great works...

![Screenshot_20210703_011758](https://user-images.githubusercontent.com/6520357/124343517-6031e680-dba2-11eb-9119-2659964750f1.png)

... I enter the login, password, click on _Log in_ and I'm redirected to this URL...

`https://console-openshift-console.apps.mbr.mydomain.net/error?error=invalid_code&error_type=auth`

... which in turn redirects to this (old domain)...

`https://oauth-openshift.apps.mbr.okd.local/oauth/authorize?client_id=console&redirect_uri=https%3A%2F%2Fconsole-openshift-console.apps.mbr.mydomain.net%2Fauth%2Fcallback&response_type=code&scope=user%3Afull&state=da4a1dd0`

![Screenshot_20210703_012317](https://user-images.githubusercontent.com/6520357/124343530-80fa3c00-dba2-11eb-8d51-903b2cb9a161.png)

-----------------------------------------

**CONCLUSION:**

Even though we successfully updated the routes as exposed, we started to have **problems with the login** and the **URLs of the _Web Console_ that are not updated** to the new domain.

-----------------------------------------

-----------------------------------------

**QUESTION:**

Does anyone have any idea how we can solve the mentioned problems?

-----------------------------------------

**IMPORTANT:**
 I - We've already tried this procedure [Customizing the web console URL](https://docs.openshift.com/container-platform/4.6/web_console/customizing-the-web-console.html#customizing-the-web-console-url_customizing-web-console) , but it doesn't update all routes and does not update URLs in the _Web Console_ and also has the same login issues above;
 II - I also tried this procedure [Creating a route through an Ingress object](https://docs.okd.io/latest/networking/routes/route-configuration.html#nw-ingress-creating-a-route-via-an-ingress_route-configuration) , but equally it doesn't seem to me something that will solve the problems pointed out.

**Comment by eduardolucioac** [comment]
We found **SOLUTIONS** for the problems mentioned...

> Even though we successfully updated the routes as exposed, we started to have **problems with the login** and the **URLs of the _Web Console_ that are not updated** to the new domain.

I - **problems with the login**:

It is not necessary to modify the ingress via the `oc edit ingress.config.openshift.io` command. Instead we will create a new route for the _oauth-openshift_ resource.

Open for editing (but we won't edit it) the route configuration for the _oauth-openshift_ resource...

```
oc edit -n openshift-authentication route.route.openshift.io/oauth-openshift
```

... , copy its content and edit the following properties according to your reality modifing from something like this...

```
[...]
metadata:
[...]
  name: oauth-openshift
[...]
spec:
  host: oauth-openshift.apps.mbr.okd.local
[...]
status:
[...]
    host: oauth-openshift.apps.mbr.okd.local
[...]
```

... to something like this...

```
[...]
metadata:
[...]
  name: oauth-openshift-mydomain-net
[...]
spec:
  host: oauth-openshift.apps.mbr.mydomain.net
[...]
status:
[...]
    host: oauth-openshift.apps.mbr.mydomain.net
[...]
```

... and create the new route...

```
cat <<EOF | oc create -f -
---
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  creationTimestamp: "2021-06-25T18:45:29Z"
  labels:
    app: oauth-openshift
  name: oauth-openshift-mydomain-net
  namespace: openshift-authentication
  resourceVersion: "6304382"
  uid: 44d8fdcd-42eb-40f8-8312-789dc4bb84df
spec:
  host: oauth-openshift.apps.mbr.mydomain.net
  port:
    targetPort: 6443
  tls:
    insecureEdgeTerminationPolicy: Redirect
    termination: passthrough
  to:
    kind: Service
    name: oauth-openshift
    weight: 100
  wildcardPolicy: None
status:
  ingress:
  - conditions:
    - lastTransitionTime: "2021-07-10T02:46:57Z"
      status: "True"
      type: Admitted
    host: oauth-openshift.apps.mbr.mydomain.net
    routerCanonicalHostname: apps.mbr.okd.local
    routerName: default
    wildcardPolicy: None
EOF
```

Done! We can now login using the URL `oauth-openshift.apps.mbr.mydomain.net` (mydomain.net domain)!

II - **URLs of the _Web Console_ that are not updated**

When we adjust the routes...

```
oc edit -n openshift-console route.route.openshift.io/console
oc edit -n openshift-console route.route.openshift.io/downloads
oc edit -n openshift-ingress-canary route.route.openshift.io/canary
oc edit -n openshift-monitoring route.route.openshift.io/alertmanager-main
oc edit -n openshift-monitoring route.route.openshift.io/grafana
oc edit -n openshift-monitoring route.route.openshift.io/prometheus-k8s
oc edit -n openshift-monitoring route.route.openshift.io/thanos-querier
```

... the URLs for these resources will be automatically updated in the Web Console.

--------------------------------------------

**NEW PROBLEM:**

The Web Console and its components ("alertmanager", "grafana", "prometheus", etc) always redirect to this URL...

```
https://oauth-openshift.apps.mbr.okd.local
```

... when login is required.

**NOTE:** The "okd.local" is the cluster's base domain (installation).

Is it possible to configure them to calling another URL? Like this URL below?

```
https://oauth-openshift.apps.mbr.domain.my
```

**Thanks! =D**

---------------------

**PLUS:** As I said in other spaces of our community, it is very common to have cases where the external domain (internet, WAN) of a company is different from its internal domain (private, LAN) or even cases where a company has several domains (eg: company.com , company.org, company.net...). So I say again that not being able to make web resources (especially the Web Console) work in a domain other than its internal domain is something that can become dramatic if we try to expose OpenShift (OKD) to external access (internet, WAN) a cluster already in use if its internal domain (eg: okd.local) has no external equivalent.

**Comment by vrutkovs** [comment, marked as answer]
Dupe of #716

Also, as noted in https://github.com/openshift/okd/discussions/738 changing baseDomain and thus updating oauth URLs is not supported. Please refrain from creating duplicate discussions
